src/System.Net.Primitives/src/System/Net/IPAddress.cs

```diff
@@ -608,30 +607,29 @@ public override bool Equals(object comparand)
 
         public override int GetHashCode()
         {
-            // For IPv6 addresses, we cannot simply return the integer
-            // representation as the hashcode. Instead, we calculate
-            // the hashcode from the address literals and ScopeId.
-            if (IsIPv6)
+            if (_hashCode != 0)
             {
-                if (_hashCode == 0)
-                {
-                    int hashCode = HashHelpers.Combine(HashHelpers.RandomSeed, unchecked((int)PrivateScopeId));
-                    ushort[] numbers = _numbers;
-                    for (int i = 0; i < numbers.Length; i += 2)
-                    {
-                        hashCode = HashHelpers.Combine(hashCode, numbers[i] << 16 | numbers[i + 1]);
-                    }
-
-                    _hashCode = hashCode;
-                }
-
                 return _hashCode;
             }
-            else
+
+            // For IPv4 this computes the hash for the entire Address.
+            // For IPv6 it only computes the hash for the ScopeId,
+            // so we combine it with the hash of the Address.
+            int hashCode = Marvin.ComputeHash32(
+                ref Unsafe.As<uint, byte>(ref _addressOrScopeId),
+                sizeof(uint),
+                Marvin.DefaultSeed);
+
+            if (IsIPv6)
             {
-                // For IPv4 addresses, we can simply use the integer representation.
-                return unchecked((int)PrivateAddress);
+                hashCode ^= Marvin.ComputeHash32(
```

Comment:
Is it possible to feed the scope and address into Marvin together? I'm concerned the XOR here could be attackable. For example, if it's possible to build an IPv6 with the same scope and address bytes, they'll cancel out and hash to zero.

Reply:
I don't see how. The API accepts a `ref byte` and a count.
But I guess we can instead use the scope's hash result as the seed of the address `ComputeHash32`. That way you include both values without a XOR.

Reply:
The random seeding is important (otherwise collisions become feasible). You could do something like copy to a span on the stack or maybe wrap a span around the class instance (though that makes assumptions about its layout).

Reply:
On closer inspection, wrapping a span around the class doesn't work (I missed that _numbers is an array field), so that probably leaves copying. If that really doesn't work, it might be ok to XOR the scope and Marvin.DefaultSeed though it's a little hard to reason about what that does.

Reply:
I used a `Span` over a stackalloc-ed byte array and copied the Address bytes and the ScopeId onto it before calling `Marvin`.